# <center>**Proyecto Final**</center>

## Instalar librerías

In [1]:
#instalar librerias necesarias
%pip install unidecode


[notice] A new release of pip is available: 23.2.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar librerías

In [2]:
# Importar librerias
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from unidecode import unidecode
from sklearn import preprocessing
import warnings
import sqlite3

## Consumir dataset completo desde un archivo _csv_ local

In [3]:
try:
    # Leer el archivo CSV ubicado en la ruta especificada y asigna los datos a la variable df_resultados
    df_casos_covid_colombia = pd.read_csv('COVID-19-Colombia.csv')

    # Si la lectura del archivo es exitosa, imprime un mensaje indicando que el dataset se cargó correctamente
    print("Dataset cargado correctamente")

# Si se produce un error específico de "FileNotFoundError" (archivo no encontrado), ejecuta este bloque de código
except FileNotFoundError as e:

    # Imprime un mensaje de error personalizado indicando que el archivo no se pudo encontrar, junto con los detalles específicos del error
    print(f"Error: no se encuentra el archivo. Detalles {e}")

# Si se produce cualquier otro tipo de error, ejecuta este bloque de código
except Exception as ex:

    # Imprime un mensaje de error genérico indicando que ocurrió un error inesperado, junto con los detalles específicos del error
    print(f"Error: es otro tipo de error. Detalles {ex}")

C:\Users\M&A\AppData\Local\Temp\ipykernel_13864\1298963441.py:3: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  df_casos_covid_colombia = pd.read_csv('COVID-19-Colombia.csv')


Dataset cargado correctamente


## Identificación del DataFrame

In [4]:
df_casos_covid_colombia.columns

Index(['fecha reporte web', 'ID de caso', 'Fecha de notificación',
       'Código DIVIPOLA departamento', 'Nombre departamento',
       'Código DIVIPOLA municipio', 'Nombre municipio', 'Edad',
       'Unidad de medida de edad', 'Sexo', 'Tipo de contagio',
       'Ubicación del caso', 'Estado', 'Código ISO del país',
       'Nombre del país', 'Recuperado', 'Fecha de inicio de síntomas',
       'Fecha de muerte', 'Fecha de diagnóstico', 'Fecha de recuperación',
       'Tipo de recuperación', 'Pertenencia étnica',
       'Nombre del grupo étnico'],
      dtype='object')

In [5]:
# Muestra del dataframe

pd.options.display.max_columns = None #mostrar todas las columnas

df_casos_covid_colombia.sample(4)

fecha reporte web  ID de caso Fecha de notificación  \
2838654  2020-11-10 00:00:00     1155406   2020-11-07 00:00:00   
3151026  2021-06-08 00:00:00     3604932   2021-05-21 00:00:00   
1213240  2021-05-05 00:00:00     2923255   2021-04-17 00:00:00   
2096687  2020-08-20 00:00:00      506440   2020-08-12 00:00:00   

         Código DIVIPOLA departamento Nombre departamento  \
2838654                            11              BOGOTA   
3151026                            68           SANTANDER   
1213240                            25        CUNDINAMARCA   
2096687                            20               CESAR   

         Código DIVIPOLA municipio Nombre municipio  Edad  \
2838654                      11001           BOGOTA    26   
3151026                      68001      BUCARAMANGA    92   
1213240                      25001     AGUA DE DIOS    40   
2096687                      20001       VALLEDUPAR    49   

         Unidad de medida de edad Sexo Tipo de contagio Ubicación del caso  \
2838654                         1    F      Comunitaria               Casa   
3151026                         1    F      Comunitaria               Casa   
1213240                         1    F      Relacionado               Casa   
2096687                         1    F      Comunitaria               Casa   

        Estado  Código ISO del país Nombre del país  Recuperado  \
2838654   Leve                  NaN             NaN  Recuperado   
3151026   Leve                  NaN             NaN  Recuperado   
1213240   Leve                  NaN             NaN  Recuperado   
2096687   Leve                  NaN             NaN  Recuperado   

        Fecha de inicio de síntomas Fecha de muerte Fecha de diagnóstico  \
2838654         2020-10-28 00:00:00             NaN  2020-11-08 00:00:00   
3151026         2021-05-16 00:00:00             NaN  2021-06-01 00:00:00   
1213240         2021-04-02 00:00:00             NaN  2021-04-17 00:00:00   
2096687         2020-08-12 00:00:00             NaN  2020-08-20 00:00:00   

        Fecha de recuperación Tipo de recuperación  Pertenencia étnica  \
2838654   2020-11-22 00:00:00               Tiempo                 6.0   
3151026   2021-06-09 00:00:00               Tiempo                 6.0   
1213240   2021-05-06 00:00:00               Tiempo                 6.0   
2096687   2020-09-02 00:00:00               Tiempo                 6.0   

        Nombre del grupo étnico  
2838654                     NaN  
3151026                     NaN  
1213240                     NaN  
2096687                     NaN

In [6]:
# Consulta y validación de la cantidad de filas y columnas del dataframe
num_rows, num_cols = df_casos_covid_colombia.shape

# Imprimir mensaje y valores de salidas del dataframe
print(f'El Dataframe tiene {num_rows} filas y {num_cols} columnas.')

El Dataframe tiene 6390971 filas y 23 columnas.


In [7]:
# Mostrar el tipo de dato de cada columna del DataFrame
df_casos_covid_colombia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6390971 entries, 0 to 6390970
Data columns (total 23 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   fecha reporte web             object 
 1   ID de caso                    int64  
 2   Fecha de notificación         object 
 3   Código DIVIPOLA departamento  int64  
 4   Nombre departamento           object 
 5   Código DIVIPOLA municipio     int64  
 6   Nombre municipio              object 
 7   Edad                          int64  
 8   Unidad de medida de edad      int64  
 9   Sexo                          object 
 10  Tipo de contagio              object 
 11  Ubicación del caso            object 
 12  Estado                        object 
 13  Código ISO del país           float64
 14  Nombre del país               object 
 15  Recuperado                    object 
 16  Fecha de inicio de síntomas   object 
 17  Fecha de muerte               object 
 18  Fecha de diagnóstico  

De las 23 columnas del DataFrame, 16 son de tipo `object`, 5 de tipo `int64` y 2 de tipo `float64`. Y la cantidad y el porcentaje de valores nulos o faltantes por cada columna es el siguiente:

In [8]:
# Mostrar la cantidad de valoes nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
fecha reporte web                          0               0.00
ID de caso                                 0               0.00
Fecha de notificación                      0               0.00
Código DIVIPOLA departamento               0               0.00
Nombre departamento                        0               0.00
Código DIVIPOLA municipio                  0               0.00
Nombre municipio                           0               0.00
Edad                                       0               0.00
Unidad de medida de edad                   0               0.00
Sexo                                       0               0.00
Tipo de contagio                           0               0.00
Ubicación del caso                     41268               0.65
Estado                                 41268               0.65
Código ISO del país                  6387261              99.94
Nombre del país                      6387253              99.94
Recuperado                             36327               0.57
Fecha de inicio de síntomas           505384               7.91
Fecha de muerte                      6206578              97.11
Fecha de diagnóstico                    2755               0.04
Fecha de recuperación                 182324               2.85
Tipo de recuperación                  182398               2.85
Pertenencia étnica                      2156               0.03
Nombre del grupo étnico              6307107              98.69

Las columnas con más cantidad de nulos son `Código ISO del país` y `Nombre del país` ($99.94$%), seguidas por `Nombre del grupo étnico` ($98.69$%) y `Fecha de muerte` ($97.11$%).

## Exploración del DataFrame

In [9]:
# Identificar valores únicos por columna
print("\nValores únicos por columna:")
for column in df_casos_covid_colombia.columns:
    unique_values = df_casos_covid_colombia[column].unique()
    print(f"\n{column} ({len(unique_values)} valores únicos):\n{unique_values}")


Valores únicos por columna:

fecha reporte web (872 valores únicos):
['2020-12-24 00:00:00' '2020-11-08 00:00:00' '2020-08-05 00:00:00'
 '2020-12-08 00:00:00' '2020-08-17 00:00:00' '2020-07-25 00:00:00'
 '2021-01-14 00:00:00' '2021-01-12 00:00:00' '2020-08-09 00:00:00'
 '2020-07-18 00:00:00' '2020-05-23 00:00:00' '2021-01-13 00:00:00'
 '2020-09-13 00:00:00' '2020-07-09 00:00:00' '2020-09-23 00:00:00'
 '2020-09-28 00:00:00' '2020-10-08 00:00:00' '2020-09-03 00:00:00'
 '2020-10-23 00:00:00' '2020-06-25 00:00:00' '2020-08-13 00:00:00'
 '2020-11-20 00:00:00' '2021-01-07 00:00:00' '2020-10-03 00:00:00'
 '2020-08-25 00:00:00' '2020-12-19 00:00:00' '2020-12-26 00:00:00'
 '2020-07-31 00:00:00' '2020-12-13 00:00:00' '2020-12-22 00:00:00'
 '2020-12-28 00:00:00' '2020-10-31 00:00:00' '2020-10-20 00:00:00'
 '2020-05-24 00:00:00' '2021-01-10 00:00:00' '2020-12-06 00:00:00'
 '2020-09-18 00:00:00' '2020-11-27 00:00:00' '2020-10-26 00:00:00'
 '2020-12-18 00:00:00' '2020-12-11 00:00:00' '2020-10-19 00

Se observa que la información proporcionada por las siguientes columnas es redundante y/o no aporta información relevante para nuestro proyecto:

- Fecha de reporte web
- ID del caso
- Tipo de recuperación
- Nombre del grupo étnico

En las siguientes columnas se evidencia errores en los códigos:

- Código DIVIPOLA departamento ($47001$, $8001$, $13001$), generando error tambien en el nombre de Nombre departamento.
- Código DIVIPOLA municipio

Además, se deben unificar/estandarizar los valores o categorías en las siguientes columnas:

- Nombre departamento
- Nombre municipio
- Sexo
- Ubicación del caso
- Estado
- Nombre del país
- Recuperado
- Tipo de recuperación

Adicionalmente, se detectaron errores en las columnas `Unidad de medida de edad`, pues las categorías $5$ y $4$ no forman parte del diccionario de datos, `código ISO del país`, dado que estos códigos no pueden ser de 4 cifras, y `Pertenencia étnica`, donde no deben existir valores nulos de acuerdo al diccionario de datos.

In [10]:
for columna in df_casos_covid_colombia.columns:
    print(f'\n********** Columna {columna} **********')
    print(df_casos_covid_colombia[columna].value_counts())


********** Columna fecha reporte web **********
2022-01-15 00:00:00    35576
2022-01-14 00:00:00    34924
2021-06-26 00:00:00    33593
2021-06-24 00:00:00    32997
2021-06-25 00:00:00    32732
                       ...  
2020-03-11 00:00:00        6
2020-03-13 00:00:00        5
2020-03-12 00:00:00        5
2020-03-09 00:00:00        2
2020-03-06 00:00:00        1
Name: fecha reporte web, Length: 872, dtype: int64

********** Columna ID de caso **********
1556979    1
4144546    1
4158586    1
4158585    1
4158584    1
          ..
1238778    1
700904     1
700903     1
700902     1
4990096    1
Name: ID de caso, Length: 6390971, dtype: int64

********** Columna Fecha de notificación **********
2022-01-06 00:00:00    44424
2022-01-07 00:00:00    41706
2022-01-05 00:00:00    37292
2021-06-10 00:00:00    36880
2021-06-15 00:00:00    36602
                       ...  
2020-03-08 00:00:00        2
2020-03-06 00:00:00        1
2020-03-02 00:00:00        1
2020-03-03 00:00:00        1
2020-

## Limpieza de datos

### Eliminar categorías $4$ y $5$ de la columna `Unidad de medida de edad`

In [11]:
# Eliminar filas donde 'Unidad de medida de edad' es 4 o 5
df_casos_covid_colombia = df_casos_covid_colombia[~df_casos_covid_colombia['Unidad de medida de edad'].isin([4, 5])]

### Corrección de errores en las columnas `Código DIVIPOLA departamento` y `Código DIVIPOLA municipio`

Inicialmente se realizará un reemplazo de los códigos que no corresponden a cada columna, posteriormente se cargará un dataset que contenga los códigos DIVIPOLA de departamentos y municipios para hacer una unión con el dataframe principal (df_casos_covid_colombia).

In [12]:
# Códigos DIVIPOLA a reemplazar en la columna 'Código DIVIPOLA departamento'
reemplazar_codigos_departamento = {47001:47, 8001:8, 13001:13}

#Implementar la función .replace para realizar el reemplazo
df_casos_covid_colombia['Código DIVIPOLA departamento'].replace(reemplazar_codigos_departamento, inplace=True)
df_casos_covid_colombia['Código DIVIPOLA departamento'].unique()

array([76,  5, 68, 54, 17, 11, 25, 18, 63, 66, 85, 52, 41, 20, 50, 47, 73,
       19,  8, 70, 13, 44, 23, 81, 15, 88, 27, 95, 99, 86, 97, 94, 91],
      dtype=int64)

In [13]:
# Cargar dataset que contiene los códigos DIVIPOLA de departamentos y municipios

try:
    # Leer el archivo CSV ubicado en la ruta especificada y asigna los datos a la variable df_resultados
    df_codigos_DIVIPOLA = pd.read_csv('Codigos_DIVIPOLA.csv', dtype=str)

    # Si la lectura del archivo es exitosa, imprime un mensaje indicando que el dataset se cargó correctamente
    print("Dataset cargado correctamente")

# Si se produce un error específico de "FileNotFoundError" (archivo no encontrado), ejecuta este bloque de código
except FileNotFoundError as e:

    # Imprime un mensaje de error personalizado indicando que el archivo no se pudo encontrar, junto con los detalles específicos del error
    print(f"Error: no se encuentra el archivo. Detalles {e}")

# Si se produce cualquier otro tipo de error, ejecuta este bloque de código
except Exception as ex:

    # Imprime un mensaje de error genérico indicando que ocurrió un error inesperado, junto con los detalles específicos del error
    print(f"Error: es otro tipo de error. Detalles {ex}")

Dataset cargado correctamente


In [14]:
# Muestra del DataFrame 'df_codigos_DIVIPOLA'
df_codigos_DIVIPOLA.head(5)

Código Departamento Código Municipio Nombre Departamento Nombre Municipio  \
0                   5             5001           ANTIOQUIA         MEDELLÍN   
1                   5             5002           ANTIOQUIA        ABEJORRAL   
2                   5             5004           ANTIOQUIA         ABRIAQUÍ   
3                   5             5021           ANTIOQUIA       ALEJANDRÍA   
4                   5             5030           ANTIOQUIA            AMAGÁ   

  Tipo: Municipio / Isla / Área no municipalizada  
0                                       Municipio  
1                                       Municipio  
2                                       Municipio  
3                                       Municipio  
4                                       Municipio

In [15]:
# Información de las columnas del DataFrame 'df_codigos_DIVIPOLA'
df_codigos_DIVIPOLA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1129 entries, 0 to 1128
Data columns (total 5 columns):
 #   Column                                           Non-Null Count  Dtype 
---  ------                                           --------------  ----- 
 0   Código Departamento                              1128 non-null   object
 1   Código Municipio                                 1123 non-null   object
 2   Nombre Departamento                              1123 non-null   object
 3   Nombre Municipio                                 1123 non-null   object
 4   Tipo: Municipio / Isla / Área no municipalizada  1121 non-null   object
dtypes: object(5)
memory usage: 44.2+ KB


In [16]:
# Unir dataframes

# Selecciona las primeras 1123 filas y todas las columnas del DataFrame df_codigos_DIVIPOLA
df_codigos_DIVIPOLA = df_codigos_DIVIPOLA.iloc[:1123,:]
# Convierte las columnas 'Código Departamento' y 'Código Municipio' a tipo entero de 64 bits
df_codigos_DIVIPOLA[['Código Departamento', 'Código Municipio']] = df_codigos_DIVIPOLA[['Código Departamento', 'Código Municipio']].astype('int64')


# Crea un nuevo DataFrame que contiene solo las columnas 'Código Departamento' y 'Nombre Departamento' del DataFrame original,
# eliminando las filas duplicadas para tener una lista única de códigos y nombres de departamentos.
df_codigos_DIVIPOLA_departamento = df_codigos_DIVIPOLA[['Código Departamento', 'Nombre Departamento']].drop_duplicates()
# Realiza una combinación (merge) del DataFrame df_casos_covid_colombia con df_codigos_DIVIPOLA_departamento.
# La combinación se realiza a la izquierda (left join), utilizando 'Código DIVIPOLA departamento' del primer DataFrame
# y 'Código Departamento' del segundo DataFrame como claves de unión.
df_casos_covid_colombia = pd.merge(df_casos_covid_colombia, df_codigos_DIVIPOLA_departamento, how='left', 
                                   left_on='Código DIVIPOLA departamento', right_on='Código Departamento')


# Mismo procedimiento para los códigos DIVIPOLA de los municipios
df_codigos_DIVIPOLA_municipio = df_codigos_DIVIPOLA[['Código Municipio', 'Nombre Municipio']].drop_duplicates()
df_casos_covid_colombia = pd.merge(df_casos_covid_colombia, df_codigos_DIVIPOLA_municipio, how='left', 
                                   left_on='Código DIVIPOLA municipio', right_on='Código Municipio')

### Corrección de errores en la columna `Código ISO del país`

Al igual que en las columnas anteriores, inicialmente se realizará un reemplazo de los códigos ISO que no corresponden a un país, posteriormente se cargará un dataset que contenga dichos códigos para hacer una unión con el dataframe principal (df_casos_covid_colombia).

In [17]:
# Códigos ISO a reemplazar
reemplazar_codigos_pais = {1724:724, 1249:840, 1152:152, 1589:604, 1380:380, 1000:528, 1580:591, 
                           1239:218, 77:76, 531:528, 78:76, 85:76, 81:76, 87:76, 79:76, 80:76}

#Implementar la función .replace para realizar el reemplazo
df_casos_covid_colombia['Código ISO del país'].replace(reemplazar_codigos_pais, inplace=True)
df_casos_covid_colombia['Código ISO del país'].unique()

array([  nan,  484.,  214.,  276.,  250.,  840.,  218.,  620.,   76.,
        222.,  591.,  784.,   51.,  212.,  578.,  604.,  724.,  360.,
        862.,  152.,  780.,  558.,  850.,   32.,  380.,  528.,  630.,
        818.,  643.,  124.,  320.,  682.,  792.,  654.,  112.,  826.,
        340.,  376.,  832.,   36.,  188.,  600.,  804.,  533.,  388.,
        422.,   68.,  670.,   20.,  140.,  858., 1528.,  192.,  752.,
        120.,  660.,  312.,    8.,  404.,  530.,   56.,  504.,  372.,
        300.,  270.,  756.,  634.,  566.,   84.,   86.,  174.,  191.,
        308.,  760.,    4.,  258.,  410.,  499.,  156.,   70.,  652.,
        178.,  203.,  332.])

In [18]:
# Cargar dataset que contiene los códigos ISO de los países

try:
    # Leer el archivo CSV ubicado en la ruta especificada y asigna los datos a la variable df_resultados
    df_codigos_ISO = pd.read_csv('codigo-ISO-pais.csv', dtype=str)

    # Si la lectura del archivo es exitosa, imprime un mensaje indicando que el dataset se cargó correctamente
    print("Dataset cargado correctamente")

# Si se produce un error específico de "FileNotFoundError" (archivo no encontrado), ejecuta este bloque de código
except FileNotFoundError as e:

    # Imprime un mensaje de error personalizado indicando que el archivo no se pudo encontrar, junto con los detalles específicos del error
    print(f"Error: no se encuentra el archivo. Detalles {e}")

# Si se produce cualquier otro tipo de error, ejecuta este bloque de código
except Exception as ex:

    # Imprime un mensaje de error genérico indicando que ocurrió un error inesperado, junto con los detalles específicos del error
    print(f"Error: es otro tipo de error. Detalles {ex}")

Dataset cargado correctamente


In [19]:
# Muestra del DataFrame 'df_codigos_DIVIPOLA'
df_codigos_ISO.head(5)

NOMBRE_PAIS CODIGO_ISO_PAIS
0                 Andorra              20
1  Emiratos Árabes Unidos             784
2              Afganistán               4
3       Antigua y Barbuda              28
4                 Anguila             660

In [20]:
#Información de las columnas del DataFrame
df_codigos_ISO.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   NOMBRE_PAIS      240 non-null    object
 1   CODIGO_ISO_PAIS  240 non-null    object
dtypes: object(2)
memory usage: 3.9+ KB


In [21]:
# Unir dataframes

# Selecciona las primeras 240 filas y todas las columnas del DataFrame df_codigos_ISO.
df_codigos_ISO = df_codigos_ISO.iloc[:239,:]

# Convierte las columnas 'CODIGO_ISO_PAIS' a tipo entero de 64 bits.
df_codigos_ISO['CODIGO_ISO_PAIS'] = df_codigos_ISO['CODIGO_ISO_PAIS'].astype('int64')

# Crea un nuevo DataFrame que contiene solo las columnas 'CODIGO_ISO_PAIS' del DataFrame original,
# eliminando las filas duplicadas para tener una lista única de códigos y nombres de departamentos.
df_codigos_ISO_unico = df_codigos_ISO[['CODIGO_ISO_PAIS', 'NOMBRE_PAIS']].drop_duplicates()

# Realiza una combinación (merge) del DataFrame df_casos_covid_colombia con df_codigos_ISO_unico.
# La combinación se realiza a la izquierda (left join), utilizando 'CODIGO_ISO_DEL_PAIS' del primer DataFrame
# y 'CODIGO_ISO_PAIS' del segundo DataFrame como claves de unión.
df_casos_covid_colombia = pd.merge(df_casos_covid_colombia, df_codigos_ISO_unico, how='left', 
                                   left_on='Código ISO del país', right_on='CODIGO_ISO_PAIS')

### Imputación de los valores nulos en la columna `NOMBRE_PAIS`

Como esta columna se refiere al país de donde proviene la persona infectada entonces los valores nulos indican que la persona se contagió en Colombia 

In [22]:
# Imputación de los valores nulos como 'Colombia'
df_casos_covid_colombia.loc[df_casos_covid_colombia['NOMBRE_PAIS'].isnull(), 'NOMBRE_PAIS'] = 'Colombia'

### Corrección/Imputación de los valores nulos en la columna `Pertenencia étnica`

Debido a que no deben existir valores nulos en la columna, se imputarán estos valores por el número $7$ aludiendo a una nueva categoría denominada 'No_aplica'.

In [23]:
# Reemplazar NaN con 7 en la columna
df_casos_covid_colombia['Pertenencia étnica'] = df_casos_covid_colombia['Pertenencia étnica'].fillna(7)

# Verificar los valores únicos después de reemplazar NaN
print(df_casos_covid_colombia['Pertenencia étnica'].unique())

[6. 5. 1. 3. 2. 7.]


In [24]:
# Categorías a reemplazar en la columna 'Pertenencia étnica'
reemplazar_pertenencia_etnica = {1:'Indigena', 2:'ROM', 3:'Raizal', 4:'Palenquero', 
                                 5:'Negro', 6:'Otro', 7:'No_Aplica'}

#Implementar la función .replace para realizar el reemplazo
df_casos_covid_colombia['Pertenencia étnica'].replace(reemplazar_pertenencia_etnica, inplace=True)
df_casos_covid_colombia['Pertenencia étnica'].unique()

array(['Otro', 'Negro', 'Indigena', 'Raizal', 'ROM', 'No_Aplica'],
      dtype=object)

### Corrección/Imputación de los valores nulos en la columna `Unidad de medida de edad`

In [25]:
# Categorías a reemplazar en la columna 'Unidad de medida de edad'
reemplazar_unidad_medida_edad = {1:'Anios', 2:'Meses', 3:'Dias'}

#Implementar la función .replace para realizar el reemplazo
df_casos_covid_colombia['Unidad de medida de edad'].replace(reemplazar_unidad_medida_edad, inplace=True)
df_casos_covid_colombia['Unidad de medida de edad'].unique()

array(['Anios', 'Meses', 'Dias'], dtype=object)

### Cambiar el tipo de dato de las columnas con **Fechas**: `Fecha de notificación`, `Fecha de inicio de síntomas`, `Fecha de muerte`, `Fecha de diagnóstico` y `Fecha de recuperación`

In [26]:
# Columnas a modificar
columnas_a_modificar = ['Fecha de notificación', 'Fecha de inicio de síntomas', 'Fecha de muerte', 
                        'Fecha de diagnóstico', 'Fecha de recuperación']

# modificación columnas
for nombre_columna in columnas_a_modificar:
    if (nombre_columna in df_casos_covid_colombia.columns):
        # Convertir la columna de fechas a tipo datetime
        df_casos_covid_colombia[nombre_columna] = pd.to_datetime(df_casos_covid_colombia[nombre_columna])
        # Formatear la columna de fechas para que solo incluya la fecha sin tiempo
        df_casos_covid_colombia[nombre_columna]  = df_casos_covid_colombia[nombre_columna].dt.date

df_casos_covid_colombia.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6390969 entries, 0 to 6390968
Data columns (total 29 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   fecha reporte web             object 
 1   ID de caso                    int64  
 2   Fecha de notificación         object 
 3   Código DIVIPOLA departamento  int64  
 4   Nombre departamento           object 
 5   Código DIVIPOLA municipio     int64  
 6   Nombre municipio              object 
 7   Edad                          int64  
 8   Unidad de medida de edad      object 
 9   Sexo                          object 
 10  Tipo de contagio              object 
 11  Ubicación del caso            object 
 12  Estado                        object 
 13  Código ISO del país           float64
 14  Nombre del país               object 
 15  Recuperado                    object 
 16  Fecha de inicio de síntomas   object 
 17  Fecha de muerte               object 
 18  Fecha de diagnóstico  

### Corrección/Imputación de los valores nulos en la columna `ubicación del caso`

In [27]:
# Reemplazar NaN 
df_casos_covid_colombia['Ubicación del caso'] = df_casos_covid_colombia['Ubicación del caso'].fillna('Desconocido')

# Verificar los valores únicos después de reemplazar NaN
print(df_casos_covid_colombia['Ubicación del caso'].unique())
# verificar conteo
df_casos_covid_colombia['Ubicación del caso'].value_counts()

['Casa' 'Fallecido' 'Desconocido' 'casa' 'CASA' 'Hospital' 'Hospital UCI']


Casa            6177079
Fallecido        143125
Desconocido       41268
casa              28328
Hospital           1056
Hospital UCI        106
CASA                  7
Name: Ubicación del caso, dtype: int64

### Corrección/Imputación de los valores nulos en la columna `Estado`

In [28]:
# Reemplazar NaN 
df_casos_covid_colombia['Estado'] = df_casos_covid_colombia['Estado'].fillna('Desconocido')

# Verificar los valores únicos después de reemplazar NaN
print(df_casos_covid_colombia['Estado'].unique())
# verificar conteo
df_casos_covid_colombia['Estado'].value_counts()

['Leve' 'Fallecido' 'Desconocido' 'leve' 'LEVE' 'Moderado' 'Grave']


Leve           6154789
Fallecido       143125
leve             50623
Desconocido      41268
Moderado          1056
Grave              106
LEVE                 2
Name: Estado, dtype: int64

### Corrección/Imputación de los valores nulos en la columna `Recuperado`

In [29]:
# Reemplazar NaN 
df_casos_covid_colombia['Recuperado'] = df_casos_covid_colombia['Recuperado'].fillna('Fallecido NO COVID')

# Verificar los valores únicos después de reemplazar NaN
print(df_casos_covid_colombia['Recuperado'].unique())
# verificar conteo
df_casos_covid_colombia['Recuperado'].value_counts()

['Recuperado' 'Fallecido' 'Fallecido NO COVID' 'fallecido' 'Activo']


Recuperado            6208571
Fallecido              142830
Fallecido NO COVID      36327
Activo                   2946
fallecido                 295
Name: Recuperado, dtype: int64

### Eliminar columnas con información redundante

In [30]:
# Definir los nombres de las columnas a eliminar
columnas_a_eliminar = ['fecha reporte web', 'ID de caso', 'Código DIVIPOLA departamento', 'Nombre departamento', 
                       'Código DIVIPOLA municipio', 'Nombre municipio', 'Tipo de recuperación', 'Código ISO del país', 
                       'Nombre del país', 'Código Departamento', 'Código Municipio', 'Nombre del grupo étnico', 'CODIGO_ISO_PAIS']

# Eliminar las columnas
df_casos_covid_colombia = df_casos_covid_colombia.drop(columns = columnas_a_eliminar)

### Unificar el nombre de las columnas a mayúsculas y sin acentos.

In [31]:
for nombre_columna in df_casos_covid_colombia.columns:
    df_casos_covid_colombia = df_casos_covid_colombia.rename( columns={ nombre_columna : unidecode(nombre_columna).upper().replace(" ", "_") } )
df_casos_covid_colombia.columns

Index(['FECHA_DE_NOTIFICACION', 'EDAD', 'UNIDAD_DE_MEDIDA_DE_EDAD', 'SEXO',
       'TIPO_DE_CONTAGIO', 'UBICACION_DEL_CASO', 'ESTADO', 'RECUPERADO',
       'FECHA_DE_INICIO_DE_SINTOMAS', 'FECHA_DE_MUERTE',
       'FECHA_DE_DIAGNOSTICO', 'FECHA_DE_RECUPERACION', 'PERTENENCIA_ETNICA',
       'NOMBRE_DEPARTAMENTO', 'NOMBRE_MUNICIPIO', 'NOMBRE_PAIS'],
      dtype='object')

### Estandarizar/Unificar las categorías de las columnas

In [32]:
# Definir los nombres de las columnas a unificar
columnas_a_unificar = ['SEXO', 'UBICACION_DEL_CASO', 'ESTADO', 'RECUPERADO', 'NOMBRE_DEL_GRUPO_ETNICO',
                       'NOMBRE_DEPARTAMENTO', 'NOMBRE_MUNICIPIO', 'NOMBRE_PAIS']

for nombre_columna in columnas_a_unificar:
    if (nombre_columna in df_casos_covid_colombia.columns):
        df_casos_covid_colombia[nombre_columna] = df_casos_covid_colombia[nombre_columna].str.replace(' ', '_').str.title()
        df_casos_covid_colombia[nombre_columna] = df_casos_covid_colombia[nombre_columna].apply(lambda x: unidecode(x) if type(x) == str else x)
        print(f'Columna unificada: {nombre_columna}')

Columna unificada: SEXO
Columna unificada: UBICACION_DEL_CASO
Columna unificada: ESTADO
Columna unificada: RECUPERADO
Columna unificada: NOMBRE_DEPARTAMENTO
Columna unificada: NOMBRE_MUNICIPIO
Columna unificada: NOMBRE_PAIS


### Información Actual del DataFrame

In [33]:
# Consulta y validación de la cantidad de filas y columnas del dataframe
num_rows, num_cols = df_casos_covid_colombia.shape

# Imprimir mensaje y valores de salidas del dataframe
print(f'El Dataframe tiene {num_rows} filas y {num_cols} columnas.')

El Dataframe tiene 6390969 filas y 16 columnas.


In [34]:
for columna in df_casos_covid_colombia.columns:
    print(f'\n********** Columna {columna} **********')
    print(df_casos_covid_colombia[columna].value_counts())


********** Columna FECHA_DE_NOTIFICACION **********
2022-01-06    44424
2022-01-07    41706
2022-01-05    37292
2021-06-10    36880
2021-06-15    36602
              ...  
2020-03-08        2
2020-03-06        1
2020-03-02        1
2020-03-03        1
2020-03-07        1
Name: FECHA_DE_NOTIFICACION, Length: 1413, dtype: int64

********** Columna EDAD **********
30     152958
29     151462
28     151306
31     149885
27     149393
        ...  
111         6
109         5
112         4
114         3
113         1
Name: EDAD, Length: 114, dtype: int64

********** Columna UNIDAD_DE_MEDIDA_DE_EDAD **********
Anios    6357475
Meses      30732
Dias        2762
Name: UNIDAD_DE_MEDIDA_DE_EDAD, dtype: int64

********** Columna SEXO **********
F    3417093
M    2973876
Name: SEXO, dtype: int64

********** Columna TIPO_DE_CONTAGIO **********
Comunitaria    4447738
Relacionado    1939528
Importado         3703
Name: TIPO_DE_CONTAGIO, dtype: int64

********** Columna UBICACION_DEL_CASO **********


In [35]:
# Mostrar la cantidad de valoes nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
FECHA_DE_NOTIFICACION                     0               0.00
EDAD                                      0               0.00
UNIDAD_DE_MEDIDA_DE_EDAD                  0               0.00
SEXO                                      0               0.00
TIPO_DE_CONTAGIO                          0               0.00
UBICACION_DEL_CASO                        0               0.00
ESTADO                                    0               0.00
RECUPERADO                                0               0.00
FECHA_DE_INICIO_DE_SINTOMAS          505384               7.91
FECHA_DE_MUERTE                     6206576              97.11
FECHA_DE_DIAGNOSTICO                   2755               0.04
FECHA_DE_RECUPERACION                182324               2.85
PERTENENCIA_ETNICA                        0               0.00
NOMBRE_DEPARTAMENTO                       0               0.00
NOMBRE_MUNICIPIO                          0               0.00
NOMBRE_PAIS                               0               0.00

Se observa la presencia de valores nulos en las columnas con **fechas**. Para determinar el tratamiento adecuado de estas variables, se consultó el diccionario de datos y se analizaron las relaciones entre ellas. *Por ejemplo*, si existe la variable `FECHA_DE_MUERTE` y la variable `FECHA_DE_RECUPERACION` presenta valores nulos, se debe revisar esta relación para determinar el manejo adecuado de los campos nulos en ambas variables.

### Corrección/Imputación de los valores nulos en las columnas `FECHA_DE_INICIO_DE_SINTOMAS` y `FECHA_DE_DIAGNOSTICO`

In [36]:
# Cantidad de registros que tienen la misma fecha para ambas columnas
df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'][df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'] 
                                                       == df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO']].value_counts().sum()

274473

Un total de $274473$ casos tienen la misma fecha para ambas columnas, lo cual nos muestra que la **fecha de inicio de sintomas** y la **fecha diagnostico**, la cual es la fecha de confirmación por parte del laboratorio, coinciden en algunos casos.

In [37]:
# Cantidad de valores nulos en 'FECHA_DE_INICIO_DE_SINTOMAS' cuando 'FECHA_DE_DIAGNOSTICO' no es nula
df_casos_covid_colombia[['FECHA_DE_INICIO_DE_SINTOMAS', 'FECHA_DE_DIAGNOSTICO']][df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'].isnull() &
                                                                                 df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO'].notnull()]

FECHA_DE_INICIO_DE_SINTOMAS FECHA_DE_DIAGNOSTICO
216                             NaT           2021-01-12
229                             NaT           2021-01-13
230                             NaT           2021-01-13
231                             NaT           2021-01-13
235                             NaT           2021-01-13
...                             ...                  ...
6390552                         NaT           2021-09-09
6390576                         NaT           2021-09-28
6390770                         NaT           2021-11-02
6390808                         NaT           2021-11-02
6390922                         NaT           2021-10-22

[505282 rows x 2 columns]

Además, se observa que del total de valores nulos ($505384$) de la columna `FECHA_DE_INICIO_DE_SINTOMAS`, $505282$ registros presentan valores en la variable `FECHA_DE_DIAGNOSTICO`. Dado que esta última representa la fecha de confirmación por laboratorio, se asumirá que la prueba se realizó por sospecha el mismo día. Por lo tanto, se procederá a llenar los valores nulos de `FECHA_DE_INICIO_DE_SINTOMAS` con las fechas presentes en la variable `FECHA_DE_DIAGNOSTICO`.

In [38]:
# Llenar los valores nulos de `FECHA_DE_INICIO_DE_SINTOMAS` con las fechas presentes en la variable `FECHA_DE_DIAGNOSTICO`
df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'] = df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'].fillna(df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO'])

Para el resto de registros ($102$), que son nulos para ambas columnas, se toma la decisión de elimnarlas dado que representa apenas el $0.001\%$ de los datos.

In [39]:
# Filtrar el DataFrame para encontrar las filas donde ambas columnas son nulas
eliminar_filas_na = df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'].isnull() & df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO'].isnull()

# Eliminar esas filas del DataFrame
df_casos_covid_colombia = df_casos_covid_colombia[~eliminar_filas_na]


Ahora verificaremos de nuevo los datos nulos del dataframe completo con los cambios realizados

In [40]:
# Mostrar la cantidad de valoes nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
FECHA_DE_NOTIFICACION                     0               0.00
EDAD                                      0               0.00
UNIDAD_DE_MEDIDA_DE_EDAD                  0               0.00
SEXO                                      0               0.00
TIPO_DE_CONTAGIO                          0               0.00
UBICACION_DEL_CASO                        0               0.00
ESTADO                                    0               0.00
RECUPERADO                                0               0.00
FECHA_DE_INICIO_DE_SINTOMAS               0               0.00
FECHA_DE_MUERTE                     6206474              97.11
FECHA_DE_DIAGNOSTICO                   2653               0.04
FECHA_DE_RECUPERACION                182324               2.85
PERTENENCIA_ETNICA                        0               0.00
NOMBRE_DEPARTAMENTO                       0               0.00
NOMBRE_MUNICIPIO                          0               0.00
NOMBRE_PAIS                               0               0.00

Ahora, para los registros nulos que tiene la columna `FECHA_DE_DIAGNOSTICO` se tomarán las fechas registradas como inicio de sintomas que las proporciona la columna `FECHA_DE_INICIO_DE_SINTOMAS`

In [41]:
df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO'] = df_casos_covid_colombia['FECHA_DE_DIAGNOSTICO'].fillna(df_casos_covid_colombia['FECHA_DE_INICIO_DE_SINTOMAS'])

# Mostrar la cantidad de valores nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
FECHA_DE_NOTIFICACION                     0               0.00
EDAD                                      0               0.00
UNIDAD_DE_MEDIDA_DE_EDAD                  0               0.00
SEXO                                      0               0.00
TIPO_DE_CONTAGIO                          0               0.00
UBICACION_DEL_CASO                        0               0.00
ESTADO                                    0               0.00
RECUPERADO                                0               0.00
FECHA_DE_INICIO_DE_SINTOMAS               0               0.00
FECHA_DE_MUERTE                     6206474              97.11
FECHA_DE_DIAGNOSTICO                      0               0.00
FECHA_DE_RECUPERACION                182324               2.85
PERTENENCIA_ETNICA                        0               0.00
NOMBRE_DEPARTAMENTO                       0               0.00
NOMBRE_MUNICIPIO                          0               0.00
NOMBRE_PAIS                               0               0.00

### Análisis y corrección/imputación de los registros en las columnas `FECHA_DE_MUERTE`, `FECHA_DE_RECUPERACION`, `RECUPERADO`

Dada la conexión existente entre estas dos variables miraremos para cuántos registros no se tiene información de ninguna de las dos

In [42]:
df_casos_covid_colombia[df_casos_covid_colombia['FECHA_DE_MUERTE'].isnull() & 
                        df_casos_covid_colombia['FECHA_DE_RECUPERACION'].isnull()].shape[0]

2946

Debido a que es posible que para estos casos no se tenga un seguimiento, pues solo se sabe que tuvieron síntomas y un diagnóstico pero no lo qué sucedió con el caso, y además representan un $0.046\%$ de los datos actuales se procede a eliminar dichos registros. 

In [43]:
# Filtrar el DataFrame para encontrar las filas donde ambas columnas son nulas
eliminar_filas_na = df_casos_covid_colombia['FECHA_DE_MUERTE'].isnull() & df_casos_covid_colombia['FECHA_DE_RECUPERACION'].isnull()

# Eliminar esas filas del DataFrame
df_casos_covid_colombia = df_casos_covid_colombia[~eliminar_filas_na]

Volvemos a verificar los datos nulos del dataframe completo con los cambios realizados

In [44]:
# Mostrar la cantidad de valoes nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
FECHA_DE_NOTIFICACION                     0               0.00
EDAD                                      0               0.00
UNIDAD_DE_MEDIDA_DE_EDAD                  0               0.00
SEXO                                      0               0.00
TIPO_DE_CONTAGIO                          0               0.00
UBICACION_DEL_CASO                        0               0.00
ESTADO                                    0               0.00
RECUPERADO                                0               0.00
FECHA_DE_INICIO_DE_SINTOMAS               0               0.00
FECHA_DE_MUERTE                     6203528              97.11
FECHA_DE_DIAGNOSTICO                      0               0.00
FECHA_DE_RECUPERACION                179378               2.81
PERTENENCIA_ETNICA                        0               0.00
NOMBRE_DEPARTAMENTO                       0               0.00
NOMBRE_MUNICIPIO                          0               0.00
NOMBRE_PAIS                               0               0.00

Ahora miraremos la conexión existente entre `FECHA_DE_MUERTE` y `FECHA_DE_RECUPERACION`

In [45]:
cant_fechaMuerteNulos_fechaRecupNoNulos = df_casos_covid_colombia[df_casos_covid_colombia['FECHA_DE_MUERTE'].isnull() & 
                                                                  df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()].shape[0]
print(f'Cantidad de registros FECHA_DE_MUERTE (nulos) y FECHA_DE_RECUPERACION (no nulos): {cant_fechaMuerteNulos_fechaRecupNoNulos}')

cant_fechaMuerteNoNulos_fechaRecupNulos = df_casos_covid_colombia[df_casos_covid_colombia['FECHA_DE_MUERTE'].notnull() & 
                                                                  df_casos_covid_colombia['FECHA_DE_RECUPERACION'].isnull()].shape[0]
print(f'Cantidad de registros FECHA_DE_MUERTE (no nulos) y FECHA_DE_RECUPERACION (nulos): {cant_fechaMuerteNoNulos_fechaRecupNulos}')

cant_fechaMuerteNoNulos_fechaRecupNoNulos = df_casos_covid_colombia[df_casos_covid_colombia['FECHA_DE_MUERTE'].notnull() & 
                                                                  df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()].shape[0]
print(f'Cantidad de registros FECHA_DE_MUERTE (no nulos) y FECHA_DE_RECUPERACION (no nulos): {cant_fechaMuerteNoNulos_fechaRecupNoNulos}')

Cantidad de registros FECHA_DE_MUERTE (nulos) y FECHA_DE_RECUPERACION (no nulos): 6203528
Cantidad de registros FECHA_DE_MUERTE (no nulos) y FECHA_DE_RECUPERACION (nulos): 179378
Cantidad de registros FECHA_DE_MUERTE (no nulos) y FECHA_DE_RECUPERACION (no nulos): 5015


Observamos que del total de registros, que son ($6387921$), existen $6203528$ registros donde no se tiene `FECHA_DE_MUERTE` pero sí `FECHA_DE_RECUPERACION`, $179378$ registros donde no se tiene `FECHA_DE_RECUPERACION` pero sí `FECHA_DE_MUERTE` y $5015$ registros donde ambas columnas existen (¡¡¡un caso particular!!!) que se verifica nuevamente pero mostrando la diferencia de días entre estas dos columnas, descartando así que `FECHA_DE_MUERTE` sea menor que `FECHA_DE_RECUPERACION`

In [46]:
diferencia_FecMuerte_FecRecup = df_casos_covid_colombia[['FECHA_DE_MUERTE', 
                                                         'FECHA_DE_RECUPERACION']][df_casos_covid_colombia['FECHA_DE_MUERTE'].notnull() & 
                                                                                   df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()]

diferencia_FecMuerte_FecRecup['DIFERENCIA'] = diferencia_FecMuerte_FecRecup['FECHA_DE_MUERTE'] - diferencia_FecMuerte_FecRecup['FECHA_DE_RECUPERACION']

diferencia_FecMuerte_FecRecup['DIFERENCIA'].sort_values()

2064855     0 days
21555       0 days
4775359     0 days
30016       0 days
4849400     0 days
            ...   
6275600   515 days
5748098   515 days
5777540   517 days
6129181   521 days
4913189   527 days
Name: DIFERENCIA, Length: 5015, dtype: timedelta64[ns]

Lo anterior corrobora que `FECHA_DE_MUERTE` es igual ($0$ días) o mayor que `FECHA_DE_RECUPERACION`. Seguidamente, se debe verificar entonces que estos $5015$ registros aparezcan como *Fallecido* en la columna `RECUPERADO`

In [47]:
df_casos_covid_colombia['RECUPERADO'][df_casos_covid_colombia['FECHA_DE_MUERTE'].notnull() & 
                                      df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()].value_counts()

Recuperado            4941
Fallecido_No_Covid      74
Name: RECUPERADO, dtype: int64

Se observa que no es el caso para $4941$ registros, por lo tanto, se procede a imputar dichos valores

In [48]:
df_casos_covid_colombia.loc[(df_casos_covid_colombia['FECHA_DE_MUERTE'].notnull()) & 
                            (df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()) & 
                            (df_casos_covid_colombia['RECUPERADO'] == 'Recuperado'), 'RECUPERADO'] = 'Fallecido'

In [49]:
df_casos_covid_colombia['RECUPERADO'][df_casos_covid_colombia['FECHA_DE_RECUPERACION'].notnull()].value_counts()

Recuperado            6203528
Fallecido                4941
Fallecido_No_Covid         74
Name: RECUPERADO, dtype: int64

## Identificación del DataFrame final

In [50]:
df_casos_covid_colombia.head(3)

FECHA_DE_NOTIFICACION  EDAD UNIDAD_DE_MEDIDA_DE_EDAD SEXO TIPO_DE_CONTAGIO  \
0            2020-12-22    67                    Anios    F      Comunitaria   
1            2020-12-19    66                    Anios    F      Comunitaria   
2            2020-12-19    68                    Anios    F      Comunitaria   

  UBICACION_DEL_CASO ESTADO  RECUPERADO FECHA_DE_INICIO_DE_SINTOMAS  \
0               Casa   Leve  Recuperado                  2020-12-21   
1               Casa   Leve  Recuperado                  2020-12-07   
2               Casa   Leve  Recuperado                  2020-12-18   

  FECHA_DE_MUERTE FECHA_DE_DIAGNOSTICO FECHA_DE_RECUPERACION  \
0             NaT           2020-12-23            2021-01-04   
1             NaT           2020-12-23            2020-12-25   
2             NaT           2020-12-22            2021-01-01   

  PERTENENCIA_ETNICA NOMBRE_DEPARTAMENTO NOMBRE_MUNICIPIO NOMBRE_PAIS  
0               Otro     Valle_Del_Cauca             Cali    Colombia  
1               Otro     Valle_Del_Cauca             Cali    Colombia  
2               Otro     Valle_Del_Cauca             Cali    Colombia

In [51]:
# Consulta y validación de la cantidad de filas y columnas del dataframe
num_rows, num_cols = df_casos_covid_colombia.shape

# Imprimir mensaje y valores de salidas del dataframe
print(f'El Dataframe final tiene {num_rows} filas y {num_cols} columnas.')

El Dataframe final tiene 6387921 filas y 16 columnas.


In [52]:
# Mostrar la cantidad de valores nulos por cada columna
cantidad_nulos = df_casos_covid_colombia.isnull().sum()
porcentaje_nulos = df_casos_covid_colombia.isnull().sum() * 100 / len(df_casos_covid_colombia)
df_nulos = pd.DataFrame({'Cantidad_Nulos': cantidad_nulos,
                         'Porcentajes_Nulos': porcentaje_nulos.round(2)})
df_nulos

Cantidad_Nulos  Porcentajes_Nulos
FECHA_DE_NOTIFICACION                     0               0.00
EDAD                                      0               0.00
UNIDAD_DE_MEDIDA_DE_EDAD                  0               0.00
SEXO                                      0               0.00
TIPO_DE_CONTAGIO                          0               0.00
UBICACION_DEL_CASO                        0               0.00
ESTADO                                    0               0.00
RECUPERADO                                0               0.00
FECHA_DE_INICIO_DE_SINTOMAS               0               0.00
FECHA_DE_MUERTE                     6203528              97.11
FECHA_DE_DIAGNOSTICO                      0               0.00
FECHA_DE_RECUPERACION                179378               2.81
PERTENENCIA_ETNICA                        0               0.00
NOMBRE_DEPARTAMENTO                       0               0.00
NOMBRE_MUNICIPIO                          0               0.00
NOMBRE_PAIS                               0               0.00

In [53]:
# Cantidad de registros por categorías en cada columna
for columna in df_casos_covid_colombia.columns:
    print(f'\n********** Columna {columna} **********')
    print(df_casos_covid_colombia[columna].value_counts())


********** Columna FECHA_DE_NOTIFICACION **********
2022-01-06    44424
2022-01-07    41706
2022-01-05    37292
2021-06-10    36880
2021-06-15    36602
              ...  
2020-03-08        2
2020-03-06        1
2020-03-07        1
2020-03-03        1
2020-03-02        1
Name: FECHA_DE_NOTIFICACION, Length: 1405, dtype: int64

********** Columna EDAD **********
30     152924
29     151419
28     151270
31     149855
27     149354
        ...  
111         6
109         5
112         4
114         3
113         1
Name: EDAD, Length: 114, dtype: int64

********** Columna UNIDAD_DE_MEDIDA_DE_EDAD **********
Anios    6354950
Meses      30271
Dias        2700
Name: UNIDAD_DE_MEDIDA_DE_EDAD, dtype: int64

********** Columna SEXO **********
F    3415376
M    2972545
Name: SEXO, dtype: int64

********** Columna TIPO_DE_CONTAGIO **********
Comunitaria    4444714
Relacionado    1939504
Importado         3703
Name: TIPO_DE_CONTAGIO, dtype: int64

********** Columna UBICACION_DEL_CASO **********


## Exportar DataFrame a un archivo csv

In [54]:
# Exportar DataFrame a un archivo csv
df_casos_covid_colombia.to_csv('casos_covid_colombia.csv', index=False)